File: kotlin_codeql/src/main/java/org/example/kotlin_codeql/Security/CWE/CWE-020/ExternalAPITaintStepExample.kt
// {fact rule=audit-log-file-deletion@v1.0 defects=unknown}
import jakarta.servlet.ServletException
import jakarta.servlet.http.HttpServlet
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import java.io.IOException
import kotlin.Throws

class SQLInjection : HttpServlet() {
    @Throws(ServletException::class, IOException::class)
    override fun doGet(request: HttpServletRequest, response: HttpServletResponse?) {
        val sqlQueryBuilder = StringBuilder()
        sqlQueryBuilder.append("SELECT * FROM user WHERE user_id='")
        sqlQueryBuilder.append(request.getParameter("user_id"))
        sqlQueryBuilder.append("'")

        // ...
    }
}
// {/fact}

File: cpp_codeql/Security/CWE/CWE-020/ExternalAPITaintStepExample.cpp
// {fact rule=improper-input-validation@v1.0 defects=1}
#include <cstdio>
#include <cstring>

namespace {
    void do_get(FILE* request, FILE* response) {
      char user_id[1024];
      fgets(user_id, 1024, request);

      char buffer[1024];
      strcat(buffer, "SELECT * FROM user WHERE user_id='");
      strcat(buffer, user_id);
      strcat(buffer, "'");

      // ...
    }
    // {/fact}
}
